gloo/transport/tcp/tls/pair.cc

```diff
@@ -0,0 +1,386 @@
+        GLOO_ENFORCE(err != SSL_ERROR_WANT_WRITE);
+
+        if (err == SSL_ERROR_WANT_READ) {
+          return false;
+        }
+
+        if (err == SSL_ERROR_ZERO_RETURN) {
+          if (errno == ECONNRESET) {
+            if (!sync_) {
+              return false;
+            }
+          }
+        }
+
+        if (err == SSL_ERROR_SYSCALL) {
+          if (errno == EPIPE) {
+            if (!sync_) {
+              return false;
+            }
+          }
+          if (errno == ECONNRESET) {
+            if (!sync_) {
+              return false; // Exit if Connection reset by peer on read
+            }
+          }
+        }
+
+        // Unexpected error
+        signalException(
+            GLOO_ERROR_MSG(
+                "Read error ", peer_.str(), ": ", err, ", errno = ", strerror(errno)));
+        return false;
+      }
+      break;
+    }
+
+    // Transition to CLOSED on EOF
+    if (rv == 0) {
+      signalException(
+          GLOO_ERROR_MSG("Connection closed by peer ", peer_.str()));
+      return false;
+    }
+
+    rx_.nread += rv;
+  }
+
+  // Read completed
+  const auto opcode = rx_.getOpcode();
+  switch (opcode) {
+    case Op::SEND_BUFFER:
+      // Done sending data to pinned buffer; trigger completion.
+      rx_.buf->handleRecvCompletion();
+      break;
+    case Op::SEND_UNBOUND_BUFFER:
+      // Remote side is sending data to unbound buffer; trigger completion
+      buf->handleRecvCompletion(rank_);
+      break;
+    case Op::NOTIFY_SEND_READY:
+      // Remote side has pending send operation
+      handleRemotePendingSend(rx_);
+      break;
+    case Op::NOTIFY_RECV_READY:
+      // Remote side has pending recv operation
+      handleRemotePendingRecv(rx_);
+      break;
+  }
+
+  // Reset read operation state.
+  rx_ = Op();
+  return true;
+}
+
+void Pair::handleEvents(int events) {
+  // Try to acquire the pair's lock so the device thread (the thread
+  // that ends up calling handleEvents) can mutate the tx and rx op
+  // fields of this instance. If the lock cannot be acquired that
+  // means some other thread is trying to mutate this pair's state,
+  // which in turn might require calling into (and locking) the
+  // underlying device (for example, when the pair transitions to the
+  // CLOSED state). To avoid deadlocks, attempt to lock the pair and
+  // skip handling the events until the next tick if the lock cannot
+  // be acquired.
+  std::unique_lock<std::mutex> lock(m_, std::try_to_lock);
+  if (!lock) {
+    return;
+  }
+
+  // State must be <= CONNECTED.
+  // If state is CLOSED; this function will NOT be called. Refer to
+  // Pair::changeState and Device::unregisterDescriptor for more info.
+  GLOO_ENFORCE_LE(state_, CONNECTED);
+
+  // Exception must not be set.
+  // If exception is set, state must advance to CLOSED state.
+  GLOO_ENFORCE(ex_ == nullptr);
+
+  if (state_ == CONNECTED) {
+    if (!is_ssl_connected_) {
+      if (!is_client_) {
```

Comment:
compared to `tcp/pair.cc`'s implementation, looks like this `if` branch is the only part that differs. I wonder if we can we can come up with a solution that avoid code duplication.

Reply:
refactored, `handleReadWrite()` was extracted